FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc

```diff
@@ -0,0 +1,258 @@
+#include "catch.hpp"
+
+#include "oneapi/tbb/global_control.h"
+
+#include "FWCore/Concurrency/interface/chain_first.h"
+#include "FWCore/Concurrency/interface/FinalWaitingTask.h"
+#include "FWCore/Concurrency/interface/WaitingThreadPool.h"
+
+namespace {
+  constexpr char const* errorContext() { return "WaitingThreadPool test"; }
+}  // namespace
+
+TEST_CASE("Test WaitingThreadPool", "[edm::WaitingThreadPool") {
+  // Using parallelism 2 here because otherwise the
+  // tbb::task_arena::enqueue() in WaitingTaskWithArenaHolder will
+  // start a new TBB thread that "inherits" the name from the
+  // WaitingThreadPool thread.
+  oneapi::tbb::global_control control(oneapi::tbb::global_control::max_allowed_parallelism, 2);
+  edm::WaitingThreadPool pool;
+
+  SECTION("One async call") {
+    std::atomic<int> count{0};
+
+    oneapi::tbb::task_group group;
+    edm::FinalWaitingTask waitTask{group};
+    {
+      using namespace edm::waiting_task::chain;
+      auto h = first([&pool, &count](edm::WaitingTaskHolder h) {
+                 edm::WaitingTaskWithArenaHolder h2(std::move(h));
+                 pool.runAsync(
+                     std::move(h2), [&count]() { ++count; }, errorContext);
+               }) |
+               lastTask(edm::WaitingTaskHolder(group, &waitTask));
+      h.doneWaiting(std::exception_ptr());
+    }
+    waitTask.waitNoThrow();
+    REQUIRE(count.load() == 1);
+    REQUIRE(waitTask.done());
+    REQUIRE(not waitTask.exceptionPtr());
+  }
+
+  SECTION("Two async calls") {
+    std::atomic<int> count{0};
+
+    oneapi::tbb::task_group group;
+    edm::FinalWaitingTask waitTask{group};
+
+    {
+      using namespace edm::waiting_task::chain;
+      auto h = first([&pool, &count](edm::WaitingTaskHolder h) {
+                 edm::WaitingTaskWithArenaHolder h2(std::move(h));
+                 pool.runAsync(
```

Comment:
What is the underlying behavior you are trying to test here?

Reply:
The system works also when there are multiple `runAsync()` calls from the same task (i.e. both functions have finished at the time `FinalWaitingTask::waitNoThrow()` returns).